From 8ee033215acf4e7de7b4aa415539d82a54eadf64 Mon Sep 17 00:00:00 2001
From: Jose Rivera <ing.joserivera26@gmail.com>
Date: Mon, 17 May 2021 14:23:37 -0600
Subject: [PATCH] Add support for MS Team implementation

The variable ms_signaling address is now added. This variable is not stringify nor change from domain to IP.
---
 include/asterisk/res_pjsip.h     |  2 ++
 res/res_pjsip/config_transport.c | 32 ++++++++++++++++++++++++++++++++
 res/res_pjsip_nat.c              | 19 +++++++++++++++++--
 3 files changed, 53 insertions(+), 2 deletions(-)

diff --git a/include/asterisk/res_pjsip.h b/include/asterisk/res_pjsip.h
index 2fafd5790a2..6789a230475 100644
--- a/include/asterisk/res_pjsip.h
+++ b/include/asterisk/res_pjsip.h
@@ -236,6 +236,8 @@ struct ast_sip_transport {
 		AST_STRING_FIELD(external_media_address);
 		/*! Optional domain to use for messages if provided could not be found */
 		AST_STRING_FIELD(domain);
+        /*! MS Team Variable */
+        AST_STRING_FIELD(ms_signaling_address);
 		);
 	/*! Type of transport */
 	enum ast_transport type;

diff --git a/res/res_pjsip/config_transport.c b/res/res_pjsip/config_transport.c
index 2acf9a93123..5523dd161ad 100644
--- a/res/res_pjsip/config_transport.c
+++ b/res/res_pjsip/config_transport.c
@@ -1108,6 +1108,35 @@ static int privkey_file_to_str(const void *obj, const intptr_t *args, char **buf
 	return 0;
 }
 
+/*! \brief Custom handler for ms team parameter */
+static int transport_ms_signaling_address_handler(const struct aco_option *opt, struct ast_variable *var, void *obj)
+{
+	struct ast_sip_transport *transport = obj;
+    RAII_VAR(struct ast_sip_transport_state *, state, find_or_create_temporary_state(transport), ao2_cleanup);
+
+    if (!state) {
+			return -1;
+    }
+
+    if (ast_strlen_zero(var->value)) {
+        /* Ignore empty options */
+        return 0;
+    }
+
+	ast_string_field_set(transport, ms_signaling_address, var->value);
+
+    return 0;
+}
+
+static int transport_ms_signaling_address_to_str(const void *obj, const intptr_t *args, char **buf)
+{
+	const struct ast_sip_transport *transport = obj;
+
+	*buf = ast_strdup(transport->ms_signaling_address);
+
+	return 0;
+}
+
 /*! \brief Custom handler for turning a string protocol into an enum */
 static int transport_protocol_handler(const struct aco_option *opt, struct ast_variable *var, void *obj)
 {
@@ -1832,6 +1861,9 @@ int ast_sip_initialize_sorcery_transport(void)
 	ast_sorcery_object_field_register(sorcery, "transport", "domain", "", OPT_STRINGFIELD_T, 0, STRFLDSET(struct ast_sip_transport, domain));
 	ast_sorcery_object_field_register_custom(sorcery, "transport", "verify_server", "", transport_tls_bool_handler, verify_server_to_str, NULL, 0, 0);
 	ast_sorcery_object_field_register_custom(sorcery, "transport", "verify_client", "", transport_tls_bool_handler, verify_client_to_str, NULL, 0, 0);
+
+	ast_sorcery_object_field_register_custom(sorcery, "transport", "ms_signaling_address", "", transport_ms_signaling_address_handler, transport_ms_signaling_address_to_str, NULL, 0, 0);
+
	ast_sorcery_object_field_register_custom(sorcery, "transport", "require_client_cert", "", transport_tls_bool_handler, require_client_cert_to_str, NULL, 0, 0);
	ast_sorcery_object_field_register_custom(sorcery, "transport", "allow_wildcard_certs", "", transport_tls_bool_handler, allow_wildcard_certs_to_str, NULL, 0, 0);
	ast_sorcery_object_field_register_custom(sorcery, "transport", "method", "", transport_tls_method_handler, tls_method_to_str, NULL, 0, 0);

diff --git a/res/res_pjsip_nat.c b/res/res_pjsip_nat.c
index 699e5fb08be..37689c8c053 100644
--- a/res/res_pjsip_nat.c
+++ b/res/res_pjsip_nat.c
@@ -32,6 +32,7 @@
 #include "asterisk/res_pjsip_session.h"
 #include "asterisk/module.h"
 #include "asterisk/acl.h"
+#include "asterisk/strings.h"
 
 /*! URI parameter for original host/port */
 #define AST_SIP_X_AST_ORIG_HOST "x-ast-orig-host"
@@ -367,8 +368,15 @@ static pj_status_t process_nat(pjsip_tx_data *tdata)
 			pjsip_method_cmp(&cseq->method, &pjsip_invite_method) ||
 			tdata->msg->line.status.code != PJSIP_SC_MOVED_TEMPORARILY )) {
 			/* We can only rewrite the URI when one is present */
-
 			if (uri || (uri = ast_sip_get_contact_sip_uri(tdata))) {
-				pj_strdup2(tdata->pool, &uri->host, ast_sockaddr_stringify_host(&transport_state->external_signaling_address));
+				/* Do not stringify the signalling address when using MS Teams */
+				if (!ast_strlen_zero(transport->ms_signaling_address)) {
+					pj_strdup2(tdata->pool, &uri->host, transport->ms_signaling_address);
+				}else{
+					pj_strdup2(tdata->pool, &uri->host, ast_sockaddr_stringify_host(&transport_state->external_signaling_address));
+				}
 				if (transport->external_signaling_port) {
 					uri->port = transport->external_signaling_port;
 					ast_debug(4, "Re-wrote Contact URI port to %d\n", uri->port);
@@ -378,7 +386,14 @@ static pj_status_t process_nat(pjsip_tx_data *tdata)
 
 		/* Update the via header if relevant */
 		if ((tdata->msg->type == PJSIP_REQUEST_MSG) && (via || (via = pjsip_msg_find_hdr(tdata->msg, PJSIP_H_VIA, NULL)))) {
-			pj_strdup2(tdata->pool, &via->sent_by.host, ast_sockaddr_stringify_host(&transport_state->external_signaling_address));
+
+			/* Do not stringify the signalling address when using MS Teams */
+			if (!ast_strlen_zero(transport->ms_signaling_address)) {
+                pj_strdup2(tdata->pool, &via->sent_by.host, transport->ms_signaling_address);
+            }else{
+                pj_strdup2(tdata->pool, &via->sent_by.host, ast_sockaddr_stringify_host(&transport_state->external_signaling_address));
+            }
+
 			if (transport->external_signaling_port) {
 				via->sent_by.port = transport->external_signaling_port;
 			}

